### PR TITLE
feat: update package to ESM

### PR DESCRIPTION
**Pull request by andreidmt**
## Breaking Changes

- [d9ceb0d](https://github.com/asd-xiv/blocks/pull/7/commits/d9ceb0db209a730086d0b7edf19495bfb19438de) Package is now ESM only and requires node `^14.13.1 || >=16.0.0`
- [d9ceb0d](https://github.com/asd-xiv/blocks/pull/7/commits/d9ceb0db209a730086d0b7edf19495bfb19438de#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) Update all dependencies

## Changes

- [87d4e8b](https://github.com/asd-xiv/blocks/pull/7/commits/87d4e8b1b00be6bfffd2d05bcde5cc9cc2f8904b) Add [githooks](https://git-scm.com/docs/githooks) to guard from pushing :hankey: 
- [8b86f05](https://github.com/asd-xiv/blocks/pull/7/commits/8b86f05268784463559abc9e0ea3155ac2939c57) Install [commitlint](https://commitlint.js.org), [lint-staged](https://github.com/okonet/lint-staged) to guard from pushing :hankey: 
- [460d8f0](https://github.com/asd-xiv/blocks/pull/7/commits/460d8f01294885b242b02f68854b36936398dd95) Refactor CircleCI
  - Add `gitleaks` job for detecting secrets in code
  - Run `test` job in both 14 and 16 Node versions
  - Use [Cache](https://circleci.com/docs/2.0/caching/) for storing `node_modules` cross runs, use [Workspace](https://circleci.com/docs/2.0/workspaces/) for storing files cross jobs
  - Add Workflow for PR branches

## Other

- Replace [riteway](https://github.com/ericelliott/riteway) with [tape](https://github.com/substack/tape). Keeping same `given [context] should [effects]` test message format

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,8 +2,12 @@
 
 # Changelog
 
-Releases and changelog are automaticly handled by [semantic-release](https://github.com/semantic-release/semantic-release).
+Releases and changelog are automaticly handled by
```
Comment (dgilperez) [inline CHANGELOG.md:5]:
```suggestion
Releases and changelog are automatically handled by
```

File: package.json [hunk 2]
```diff
@@ -15,71 +15,86 @@
     "name": "Andrei Dumitrescu",
     "url": "https://github.com/andreidmt"
   },
-  "main": "src/index.js",
+  "contributors": [
+    {
+      "name": "David Gil",
```
Comment (dgilperez) [inline package.json:20]: 😊

File: package.json [hunk 3]
```diff
@@ -15,71 +15,86 @@
...
+  "sideEffects": false,
+  "files": [
+    "/src"
+  ],
   "scripts": {
     "----UTIL": "",
-    "setup": "rm -rf ./node_modules && npm i && npm audit fix",
-    "setup:ci": "npm ci && npm audit fix",
-    "update": "npm-check -u",
+    "prepare": "git config --local core.hooksPath .githooks",
```
Comment (dgilperez) [inline package.json:38]: Including this branch into my project's `package.json` will cause this to execute, and fail, preventing the build.

Full error: 

```
3145 timing command:install Completed in 81242ms
3146 verbose stack Error: git dep preparation failed
3146 verbose stack     at ChildProcess.<anonymous> (/userhomepath/.asdf/installs/nodejs/16.10.0/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
3146 verbose stack     at ChildProcess.emit (node:events:390:28)
3146 verbose stack     at maybeClose (node:internal/child_process:1064:16)
3146 verbose stack     at Socket.<anonymous> (node:internal/child_process:450:11)
3146 verbose stack     at Socket.emit (node:events:390:28)
3146 verbose stack     at Pipe.<anonymous> (node:net:672:12)
3147 verbose cwd /userhomepath/src/holarosita/current/rosita_cloud_ts_functions/functions
3148 verbose Darwin 20.6.0
3149 verbose argv "/userhomepath/.asdf/installs/nodejs/16.10.0/bin/node" "/userhomepath/.asdf/installs/nodejs/16.10.0/bin/npm" "i"
3150 verbose node v16.10.0
3151 verbose npm  v7.24.0
3152 error code 128
3153 error git dep preparation failed
3154 error command /userhomepath/.asdf/installs/nodejs/16.10.0/bin/node /userhomepath/.asdf/installs/nodejs/16.10.0/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/userhomepath/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
3155 error > @asd14/blocks@0.0.1 prepare
3155 error > git config --local core.hooksPath .githooks
3156 error npm WARN using --force Recommended protections disabled.
3156 error fatal: --local solo puedo ser usado dentro de un repositorio
3156 error npm ERR! code 128
3156 error npm ERR! path /userhomepath/.npm/_cacache/tmp/git-clonewGDWgU
3156 error npm ERR! command failed
3156 error npm ERR! command sh -c git config --local core.hooksPath .githooks
3156 error
3156 error npm ERR! A complete log of this run can be found in:
3156 error npm ERR!     /userhomepath/.npm/_logs/2022-05-09T19_26_42_299Z-debug.log
```

I suspect this change was intented for local dev and release purposes, not to bubble up to the projects that include this dependency, correct?

**Review by dgilperez — COMMENTED** [review]
Thanks a lot @andreidmt for working on this, and thanks for granting me collaborator permissions to the repo. I just reviewed the code, some comments. I am going to keep testing this branch locally until I can make it work in my project and then I will close my previous PR.

My PR has a few additions as well around ajv, in particular conditional schemas and ajv-keywords dependency. I'll reopen a new PR with that and switch discussion over there.

File: package.json (hunk 3 above)
Comment (dgilperez) [reply package.json:38]: Yup, I just confirmed removing that `prepare` line the including works (and apparently it builds 🥳 )

Is `prepare` some reserved word with side-effects?

Comment (andreidmt) [reply package.json:38]: Yes, [prepare](https://docs.npmjs.com/cli/v8/using-npm/scripts#life-cycle-scripts) is a special npm script that gets run after doing a `npm i` for example.

In our case, i'm locally configuring git to look at `.githooks` folder as a source for githooks - this way we can commit them in the repo and we dont need other dependencies like [husky](https://github.com/typicode/husky).

But it looks like it's messing up the consuming repo's env. Will take a look.

**Comment by andreidmt** [comment]
> My PR has a few additions as well around ajv, in particular conditional schemas and ajv-keywords dependency. I'll reopen a new PR with that and switch discussion over there.

Yes, lets keep this one scoped to ESM

File: package.json (hunk 3 above)
Comment (dgilperez) [reply package.json]: Super handy. 

I just removed the `prepare` step from this PR temporarily, so I can test this branch on a deployed environment easier.

**Comment by dgilperez** [comment]
I am happy to the new blocks from this PR works fine in my project, after making the needed changes (moving it to ESM). 

🙌🎉

Next step, testing it in a deployed staging env. Will report soon.

**Comment by dgilperez** [comment]
@andreidmt this is working for me locally - do you think it's safe to merge and release? Even though a couple of new PRs will follow.

**Comment by dgilperez** [comment]
No rush at all - I just deployed our fork to npmjs so I can deploy it (Firebase functions do not like our package.json pointing to a github dependency)

**Comment by andreidmt** [comment]
> this is working for me locally - do you think it's safe to merge and release? Even though a couple of new PRs will follow.

@dgilperez yeah, if there's something wrong, you can revert to a prev versions

**Comment by dgilperez** [comment]
🙌😊🍻

**Pull request merged**